### PR TITLE
Task3

### PR DESCRIPTION
**Pull request by kkulahin**
# React Task 3 

### Summary 

Based on your 2nd task, make an additional opportunity to create and delete cards (for cards I mean any array of objects with image and description, for example const cards = [{ id:1, category: ‘male’, imgUrl: ‘’, price: 100 }]). 

 

### Requirements 

- Each `<Card />` component could be deleted from catalogue by clicking on the cross button (❌) which can be placed somewhere on the top of each card. After clicking on the button the array of cards in your `<CardsContainer />` should be updated. 

- `<CardsContainer />` component should also additionaly render `<CardsCreationForm />` as its children. This form should contain all the required inputs for creating the card (find a way on how to use only 1 handler for all the inputs) and a submit button. By clicking on a submit button the array of cards in your `<CardsContainer />` should be updated with the new card data. 

- All the required handlers for creating/deleting cards should be in the `<CardsContainer />` component and passed down to a child components as a __*props*__. 

- Rewrite all classes components to functional components with hooks

### CODE REVIEW

File: src/components/Cards/CardsContainer.tsx [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+
+  React.useEffect(() => {
+    ApiSubject.subscribe(setCards);
+    ApiSubject.dispatch();
+    return () => {
+      ApiSubject.unsubscribe(setCards);
+    }
+  }, []);
+
+  const onDelete = (id) => {
```
Comment (MishkanJigan) [inline src/components/Cards/CardsContainer.tsx]:
```suggestion
  const onDelete = (id: number): void => {
    setCards(cards.filter(({ id: cardId }) => cardId !== id));
  }
```

File: src/components/Cards/CardsContainer.tsx [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    setCards(cards.filter(({ id: cardId }) => cardId !== id));
+  }
+
+  const onSubmit = ({ gender, title, price, imageUrl }) => {
```
Comment (MishkanJigan) [inline src/components/Cards/CardsContainer.tsx]:
```suggestion
   const onSubmit = ( cardData: ICardData ): void => {
    const id = new Date().getTime();
    setCards([...cards, { id, ...cardData }]);
  };
```

File: src/components/Cards/CardsContainer.tsx [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+    const id = new Date().getTime();
+    setCards([...cards, { id, gender, title, price, imageUrl }]);
+  };
+
+  const renderCards = React.useMemo(() => {
+    if (!cards || !cards.length) {
+      return <div>No cards yet</div>;
+    }
+    return cards.map(({ gender, id, imageUrl, price, title }) => {
```
Comment (MishkanJigan) [inline src/components/Cards/CardsContainer.tsx]: What do you think about it?
```suggestion
    return cards.map((card: ICard) => {
```

File: src/components/Cards/CardsContainer.tsx (hunk 2 above)
Comment (kkulahin) [reply src/components/Cards/CardsContainer.tsx]: Fixed
https://github.com/BlackTruth/react/blob/80c18becffd0f0491d34262d50d4f802bad3da1e/src/components/Cards/CardsContainer.tsx#L32-L35

File: src/components/Cards/CardsContainer.tsx (hunk 3 above)
Comment (kkulahin) [reply src/components/Cards/CardsContainer.tsx]: Fixed
https://github.com/BlackTruth/react/blob/80c18becffd0f0491d34262d50d4f802bad3da1e/src/components/Cards/CardsContainer.tsx#L41-L44

File: src/components/Cards/CardsContainer.tsx (hunk 1 above)
Comment (kkulahin) [reply src/components/Cards/CardsContainer.tsx]: Fixed
https://github.com/BlackTruth/react/blob/80c18becffd0f0491d34262d50d4f802bad3da1e/src/components/Cards/CardsContainer.tsx#L28-L29

**Pull request merged**